### ISSUE TITLE
µg UnifiedNlp

### ISSUE THREAD

**Issue by Mannshoch** [mvglasow/satstat#56]
Hello there. I try to get away from Google service and wish to use the "µg UnifiedNlp" apps with the possible local Providers you could link.

Could this app mayby get information about the Working of these apps?
e.g. [LocalWifiNlpBackend](https://f-droid.org/repository/browse/?fdfilter=Unified&fdid=org.fitchfamily.android.wifi_backend) or [MozillaNlpBackend](https://f-droid.org/repository/browse/?fdfilter=Unified&fdid=org.microg.nlp.backend.ichnaea)

**Comment by mvglasow** [comment]
Hi, I use µG Unified NLP myself (mostly with the OpenBmap backend), including debugging with SatStat. In fact, doing this kind of debugging is a main use case for SatStat.

µG installs itself as a location provider and can be queried over the normal location provider API, which SatStat does. 

The map view shows the data received from both location providers, with a marker indicating the location reported by the provider and a circle indicating its accuracy. By default, the red marker is for GPS and the blue one is for network (Unified NLP in this case). If additional location providers are present on the system, they are assigned random colors.

Since UnifiedNLP fuses the raw data it gets from its backends (choosing the best and sending it to the apps, with the provider marked as "network"), apps will not be able to compare locations from multiple backends. This kind of debugging would probably require modifications to UnifiedNLP itself, adding a debug interface.

The only "under-the-hood" information which apps can get is the Extras which UnifiedNLP supplies with each location. I'd have to look into what is in them, but AFAIK they contain at least a reference to the backend that supplied the location.

I do my testing on a GAPPS-free device. There is also a version of UnifiedNLP which can be installed alongside Google's location service but I have no device to test it on. Depending on how it's implemented it would either appear as a separate location provider (giving you a third marker on the map) or as the same "network" provider. In the latter case, one would, again, have to examine the Extras of each location to find out where it came from.

**Comment by breversa** [comment]
@mvglasow :
I have both GApps and GApps-free devices, all running the appropriate µG Unified NLP versions, and I'm eager to help you test. Chime in whenever you need me ! ;-)

**Comment by mvglasow** [comment]
What exactly do you need? FYI: recent builds of UnifiedNLP have some self-check features accessible through the Preferences dialog.

When running in debug mode, UnifiedNLP provides two extras with every location which may be of interest – SERVICE_BACKEND_PROVIDER and SERVICE_BACKEND_COMPONENT. Both can be used to identify the backend (and the component within the backend) which provided the location.

**Comment by breversa** [comment]
> Both can be used to identify the backend (and the component within the backend) which provided the location.

Yes, that'd be handy (to me at least), as I use several NLP backends (local wifi, GSM location, and Mozilla Location Service) at the same time and am curious to see which backend provides what location, in order to tell which one is the most precise and/or convenient.

**Comment by gdt** [comment]
I'd like to see more information too.   I'll keep my comments in this ticket limited to things which I think can work with existing interfaces.

One idea is to steal a few lines above the map and have something like one of the following

network: LocalWifi
network: Openbmap: wifi
network: Openbmap: cell

It would also be possible to assign colors statically in satstat for known providers/backends, at least for the ones that are recommended for use (openbmap * 2, localwifi, localgsm, and maybe apple and mozilla, although their network privacy properties are unclear).

**Comment by breversa** [comment]
@mvglasow :
What kind of feedback do you need on that issue ?

**Comment by mar-v-in** [comment]
@mvglasow: UnifiedNlp will give you three extras with the location:
- `SERVICE_BACKEND_PROVIDER`: a identifier string choosen by the backend. Might even be null, there is no check done on this.
- `SERVICE_BACKEND_COMPONENT`: a string representing the service component. Something like `org.openbmap.unifiedNlp.UnifiedNlpApplication/org.openbmap.unifiedNlp.services.OpenbmapNlpService`
- `OTHER_BACKEND_RESULTS`: an `ArrayList<Location>` that represents the data of backends the Fuser choose to not present as a result (because of lower accuracy or data being old).

It would be awesome of SatStat would represent all of them. I sometimes use SatStat to debug UnifiedNlp and this would solve the microg/android_packages_apps_UnifiedNlp#69. If you're interested I can also try to prepare a PR with the feature.

**Comment by n76** [comment]
Until this feature is added to SatStat (which I also use to help debug my local WiFi and GSM backends) you can look at the fuser logging in logcat to see the same other results information that mar-v-in mentions.

**Comment by mvglasow** [comment]
@mar-v-in Thanks for the input! What I currently have in mind is to use different markers for each UnifiedNLP backend, allowing users to compare the location information and accuracy data supplied by different backends.

What is the difference between `SERVICE_BACKEND_PROVIDER` and `SERVICE_BACKEND_COMPONENT`? Which one is the "right" one to identify the backend?

A PR would be very much appreciated! Be aware, though, that I intend to merge the `satstat-mapsforge0.6.1` branch into `master` soon. Be sure to base your work off that branch, as it has been heavily refactored. (I'll post here once I've completed the merge.)

**Comment by mar-v-in** [comment]
`SERVICE_BACKEND_PROVIDER` can be freely choosen by the provider, `SERVICE_BACKEND_COMPONENT` is the technical term that identifies the provider. Based on `SERVICE_BACKEND_COMPONENT` you can thus find the name also displayed inside UnifiedNlp settings, so it's probably better to use this.

**Comment by mvglasow** [comment]
The merge is through. If you want to prepare a PR, be sure to grab the latest revision of `master` and you should be good to go.

**Comment by mar-v-in** [comment]
Sorry for the delay here, I had an initial version of this running in summer, but it lacked proper UI, as in there were multiple dots on the map for each provider, but it was not visible to the user which provider matched which dot...

Currently this is not displayed anywhere for GPS and network either on the map (I know about the possibility to open the legend), but that's not a big deal because usually you can see the difference by accuracy. With more dots that's no longer easy to see, so there probably should be some sort of legend directly on the map, either drawn on the map or by using a overlaying view. What do you think about this @mvglasow ?

**Comment by azrdev** [comment]
imho android-ish way would be a deterministic color mapping, and a popup(-like) text appearing when (long-)tapping one of the markers.

**Comment by mar-v-in** [comment]
@azrdev usually the markers overlap a lot because they basically point at the same location, just with varying accuracy. Thus tapping on one of them might not be possible.

**Comment by gdt** [comment]
It could be that touching any gets a popup with the whole key.  Or that could just be a "show key" menu item.  I suspect this is aimed at pretty nerdy people who won't mind that.    Basically I will need to read the key whenever I don't remember the colors (or I suppose if I change providers and they change), and then I don't want to see it.   I'd rather have this feature even if slightly awkward than not have it and then people can make suggestions.  For me, I have only three providers, so that's 4 colors.   (Although, I would like to see separate wifi and cell fixes for openbmap and anything else that does multiple mechanisms.)

**Comment by mvglasow** [comment]
I guess a mini-legend which can be displayed and hidden as needed is probably the best way. That can be accomplished either via a menu item, floating action button or by tapping any marker.

I wouldn't display the legend permanently as it would take up valuable screen real estate.

Currently marker colors in SatStat are assigned in the following way: there are 5 or 6 predefined colors. GPS is red, network is blue. Should the device report other providers, they get assigned one of the remaining colors and the mapping is saved. That is, colors for the two standard providers are completely deterministic, and for any other providers they are stable on the same device. I'd suggest tresting the different UnifiedNLP backends as nonstandard providers: assign a different color to each newly discovered backend and store the mapping.
-- 
Sent from my Android phone with K-9 Mail. Please excuse my brevity.
